### PR TITLE
CookieParser sets Expires to DateTime.MaxValue when it's not specified

### PR DESCRIPTION
**Pull request by alnikola**
Session cookies don't have Expires property specified which normally means they get expired once the browser is closed by the user. However, it cannot be directly mapped to .Net Core HTTP implementation since there is not a user session. Thus, it was decided to treat the lifetime of a CookieContainer as "session". That means cookies without Expire and MaxAge defined will live as long as the container lives, and therefore CookieParser sets Expires to DateTime.MaxValue in such cases.

Fixes https://github.com/dotnet/corefx/issues/40934

### CODE REVIEW

File: src/libraries/Common/src/System/Net/CookieParser.cs [hunk 1]
```diff
@@ -752,6 +752,8 @@ internal Cookie Get()
                     }
                 }
             } while (!_tokenizer.Eof && !_tokenizer.EndOfCookie);
+            if (!expiresSet && cookie != null)
```
Comment (davidsh) [inline src/libraries/Common/src/System/Net/CookieParser.cs]: nit: style guide says to have a blank line before this 'if' block.

**Comment by davidsh** [comment]
>That means cookies without Expire and MaxAge defined will live as long as the container lives, and therefore CookieParser sets Expires to DateTime.MaxValue in such cases.

Setting it to `DateTime.MaxValue`  seems in conflict with the inherent default value of the Cookie.Expires property which is `DateTime.MinValue`. The default behavior of the Cookie object already was supposed to exhibit "session" behavior.

Quoting from the docs:
https://docs.microsoft.com/en-us/dotnet/api/system.net.cookie.expires?view=netcore-3.0

>Setting the Expires property to DateTime.MinValue makes this a session cookie, which is its default value.
The Kind property of Expires is used to determine if the cookie is set to DateTimeKind.Local or DateTimeKind.Utc. If the Kind property is set to DateTimeKind.Unspecified, then Utc is assumed.

So, why do we have this bug in the first place? When a cookie is created with the constructor, it already is a "session" cookie and lives as long as the object lives. I'm concerned that setting the Expires property to MaxValue isn't the right approach given that the cookie was already set as "session" lifetime using MinValue as the default for Expires property.

**Comment by alnikola** [comment]
> So, why do we have this bug in the first place?
The bug is caused by the fact that a cookie with Expires == MinValue gets instantly expired once it's added into CookieContainer.
So, the question is probably about what "session" actually means in .Net Core. As it was discussed, we agreed that session is the lifetime of a CookieContainer. However, now it's obvious that current documentation defines a different session meaning.

**Comment by davidsh** [comment]
>So, the question is probably about what "session" actually means in .Net Core. As it was discussed, we agreed that session is the lifetime of a CookieContainer. However, now it's obvious that current documentation defines a different session meaning.

@Tratcher - Do you remember why Cookie behaves this way? It seems totally broken that a newly constructed cookie is instantly expired when it is created? It seems like it is a massive bug/error in the design and documentation?

**Comment by davidsh** [comment]
@alnikola 

>As it was discussed, we agreed that session is the lifetime of a CookieContainer.

I agree with you about what is the right behavior moving forward. A cookie parsed from the wire with no 'Expires' attribute should be treated as a session cookie. And that means, it should be alive forever until the cookie object is destroyed (usually via its CookieContainer being destroyed).

**Comment by davidsh** [comment]
@dotnet/ncl @stephentoub FYI

**Comment by davidsh** [comment]
Some more thoughts on this. I think we should consider fixing Cookie class to honor the Expires property of DateTime.MinValue as a sentinel value for "infinity", i.e. "session" lifetime. That was the intention of setting it to MinValue. It seems like we need to code to explicitly check for MinValue and treat it like a sentinel value and not just a random date in the past.

This would be classified as a breaking change. But as long as we mark the PR and remember to document it as a breaking change for .NET 5, it should be ok.

**Comment by Tratcher** [comment]
MaxValue would make an easier to understand infinity sentinel, and you wouldn't have to special case it in most of the implementation. Negative values are commonly used to delete cookies via expiration. Making MinValue mean infinity is counter intuitive.

**Comment by davidsh** [comment]
>MaxValue would make an easier to understand infinity sentinel, and you wouldn't have to special case it in most of the implementation.

So, are you proposing a breaking change and changing the default value of a `new Cookie()` to have Expires property default value be DateTime.MaxValue instead of the DateTime.MinValue? That's fine too as well since we can preserve the behavior that was documented as the cookie not being expired when initially created. Of course, it's a breaking change if someone assumed that initial Cookie objects are always expired.

**Comment by alnikola** [comment]
Full .Net framework treats DateTime.**Min**Value as a sentinel value for "infinity", it has an explicit check for this in Expired property calculation. So, I propose to do it here the same way to maintain compatibility.

**Comment by alnikola** [comment]
Actually, we already have the [same logic for DateTime.MinValue in .Net Core](https://github.com/dotnet/runtime/blob/6b6d9135298500a753b3393b1c7cc4aa94617aa2/src/libraries/System.Net.Primitives/src/System/Net/Cookie.cs#L197). Moreover, I tried reproducing the incorrect behavior of an instant cookie expiration [reported in the original issue](https://github.com/dotnet/corefx/issues/40934), but **it worked properly**. I was able to successfully retrieve a cookie without Expires property from the container. I'm now wondering if the root cause of the original issue is really a broken cookie expiration. It might be something different.
The following tests **succeeds**.
```C#
        [Fact]
        public void TestExpired()
        {
            CookieContainer cc = new CookieContainer();
            Uri uri = new Uri("http://localhost");
            cc.SetCookies(uri, "name98=value98");
            Cookie actualCookie = cc.GetCookies(uri).First();
            Assert.Equal("name98", actualCookie.Name);
            Assert.Equal("value98", actualCookie.Value);
        }

**Comment by davidsh** [comment]
>Moreover, I tried reproducing the incorrect behavior of an instant cookie expiration reported in the original issue, but it worked properly.

I suggest asking the developer who opened the issue for more information.

**Comment by stephentoub** [comment]
Just catching up on the discussion here... this is currently on-hold, right?  Thanks.

**Comment by davidsh** [comment]
>Just catching up on the discussion here... this is currently on-hold, right? Thanks.

My understanding is that this PR is on hold. There is a follow up discussion that @alnikola is having on the original issue. It's possible this PR is not needed because the theory is that the existing code is already doing the right thing.

**Comment by alnikola** [comment]
Yes, the PR is on hold since I cannot reproduce the issue. Asked the issue's author for a repro.

**Comment by stephentoub** [comment]
Thanks.  Since we can't repro and thus can't test the change, and since the issue poster hasn't responded in over a month, let's close this.  We can revisit when we have something that actually needs to be fixed.

**Pull request closed without merging**